File: RDE/data/SDToolbox/MATLAB/Demo/demo_rocket_impulse.m
% Shock and Detonation Toolbox Demo Program
% 
% Computes rocket performance using quasi-one dimensional isentropic flow using 
% both frozen and equilibrium properties for a range of helium dilutions in a 
% hydrogen-oxygen mixture.  Plots impulse as a function of dilution.
%  
% ################################################################################
% Theory, numerical methods and applications are described in the following report:
% 
%     Numerical Solution Methods for Shock and Detonation Jump Conditions, S.
%     Browne, J. Ziegler, and J. E. Shepherd, GALCIT Report FM2006.006 - R3,
%     California Institute of Technology Revised September, 2018.
% 
% Please cite this report and the website if you use these routines. 
% 
% Please refer to LICENCE.txt or the above report for copyright and disclaimers.
% 
% http://shepherd.caltech.edu/EDL/PublicResources/sdt/
% 
% ################################################################################ 
% Updated September 2018
% Tested with: 
%     MATLAB 2017b and 2018a, Cantera 2.3 and 2.4
% Under these operating systems:
%     Windows 8.1, Windows 10, Linux (Debian 9)
%%
clear;clc;
% set composition and stagnation state
mech = 'Mevel2017.cti';
gas = Solution(mech);
gas1 = Solution(mech);
for helium = 1:1:40
he(helium) = (0.1*(helium-1))/(0.1*(helium-1)+1.5);
q = ['O2:0.5 H2:1.001 HE:' num2str(0.1*(helium-1))]; 
if helium == 1
    q = 'O2:0.5 H2:1.001';
end
Tc = 300;
Pc = 1.e7;
set(gas, 'T', Tc, 'P', Pc, 'X', q);
disp(['Computing chamber conditions and isentropic quasi 1-d flow for ',q,' using ',mech])
% compute equilibrium properties for constant pressure burn
equilibrate(gas,'HP');
qc = moleFractions(gas);
% Isentropic expansion from stagnation state to low pressure
Tt = temperature(gas);
St = entropy_mass(gas);
Ht = enthalpy_mass(gas);
Pt = pressure(gas);
Rt = density(gas);
ct = soundspeed_eq(gas);
gammat_eq = ct^2*Rt/Pt;
disp(['Total pressure ',num2str(Pt),' (Pa)']);
disp(['Total temperature ',num2str(Tt),' (K)']);
disp(['Total density ',num2str(Rt),' (kg/m3)']);
disp(['Total entropy ',num2str(St),' (J/kg-K)']);
disp(['Total sound speed (equilibrium) ',num2str(ct),' (m/s)']);
disp(['gamma2 (equilibrium) ',num2str(gammat_eq),' (m/s)']);
% ambient pressure for impulse computation
Pa = 0.0 ; 
disp(['Computing isentropic expansion']);
pp = Pt;
imax = 90;
for i = 1:1:imax
     pp = pp*0.95;
     sp = [St, pp];
    % compute equilibrium isentrope
    setState_SP(gas,sp);
    %equilibrate(gas,'SP','maxsteps 10000','maxiter 10000');
    equilibrate(gas,'SP');
    P(i) = pressure(gas);
    R(i) = density(gas);
    T(i) = temperature(gas);
    h_eq(i) = enthalpy_mass(gas);
    u_eq(i) = sqrt(2*(Ht-h_eq(i)));
    Isp_eq(i) = (u_eq(i)+ (P(i) - Pa)/(u_eq(i)*R(i)))/9.81 ; 
    % compute frozen impulse  (first call is so we get a good guess for
    % entropy 
    set(gas1,'Pressure',pp,'Temperature',T(i),'MoleFractions',qc);
    set(gas1,'Pressure',pp,'Entropy',St,'MoleFractions',qc);
    h_fr(i) = enthalpy_mass(gas1);
    R_fr(i) = density(gas1);
    T_fr(i) = temperature(gas1);
    u_fr(i) = sqrt(2*(Ht-h_fr(i)));
    Isp_fr(i) = (u_fr(i)+ (P(i) - Pa)/(u_fr(i)*R_fr(i)))/9.81 ; 
end
I_fr(helium) = Isp_fr(imax);
I_eq(helium) = Isp_eq(imax);
disp(['final pressure ',num2str(P(imax)),' (Pa)']);
disp(['final specific impulse (eq) ',num2str(Isp_eq(imax)),' (s)']);
disp(['final velocity (eq) ',num2str(u_eq(imax)),' (m/s)']);
disp(['final specific impulse (fr) ',num2str(Isp_fr(imax)),' (s)']);
disp(['final velocity (fr) ',num2str(u_fr(imax)),' (m/s)']);
disp('-------------------------------------------'); disp(''); disp('');
end
%
% Plot the impulses
%
figure; clf;
fontsize = 12;
plot(he(:), I_fr(:), 'r', he(:), I_eq(:), 'b');
xlabel('helium mole fraction ','FontSize',fontsize);
ylabel('Isp (s)','FontSize',fontsize);
title('Warm gas thruster performance','FontSize',fontsize);
set(gca,'FontSize',12,'LineWidth',2);
legend('Frozen','Equilibrium')
% 

File: RDE/data/SDToolbox/Python3/demo/demo_rocket_impulse.py
"""
Shock and Detonation Toolbox Demo Program

Computes rocket performance using quasi-one dimensional isentropic flow using 
both frozen and equilibrium properties for a range of helium dilutions in a 
hydrogen-oxygen mixture.  Plots impulse as a function of dilution.

################################################################################
Theory, numerical methods and applications are described in the following report:

SDToolbox - Numerical Tools for Shock and Detonation Wave Modeling,
Explosion Dynamics Laboratory, Contributors: S. Browne, J. Ziegler,
N. Bitter, B. Schmidt, J. Lawson and J. E. Shepherd, GALCIT
Technical Report FM2018.001 Revised January 2021.

Please cite this report and the website if you use these routines. 

Please refer to LICENCE.txt or the above report for copyright and disclaimers.

http://shepherd.caltech.edu/EDL/PublicResources/sdt/

################################################################################ 
Updated September 2018
Tested with: 
    Python 3.5 and 3.6, Cantera 2.3 and 2.4
Under these operating systems:
    Windows 8.1, Windows 10, Linux (Debian 9)
"""
import cantera # don't rename as ct, conflicts with a variable name in this script
from sdtoolbox.thermo import soundspeed_eq
import numpy as np
import matplotlib.pyplot as plt

# set composition and stagnation state
mech = 'Mevel2017.cti'
gas = cantera.Solution(mech)
gas1 = cantera.Solution(mech)

he = []; I_fr = []; I_eq = []
for helium in range(40):
    he.append((0.1*helium)/(0.1*helium+1.5))
    q = 'O2:0.5 H2:1.001 He:'+str(0.1*helium)
    if helium == 0:
        q = 'O2:0.5 H2:1.001'

    Tc = 300
    Pc = 1.e7
    gas.TPX = Tc,Pc,q
    print('Computing chamber conditions and isentropic quasi 1-d flow for '+q+' using '+mech)
    # compute equilibrium properties for constant pressure burn
    gas.equilibrate('HP')
    qc = gas.X
    # Isentropic expansion from stagnation state to low pressure
    Tt = gas.T
    St = gas.entropy_mass
    Ht = gas.enthalpy_mass
    Pt = gas.P
    Rt = gas.density
    ct = soundspeed_eq(gas)
    gammat_eq = ct**2*Rt/Pt
    
    print('Total pressure '+str(Pt)+' (Pa)')
    print('Total temperature '+str(Tt)+' (K)')
    print('Total density '+str(Rt)+' (kg/m3)')
    print('Total entropy '+str(St)+' (J/kg-K)')
    print('Total sound speed (equilibrium) '+str(ct)+' (m/s)')
    print('gamma2 (equilibrium) '+str(gammat_eq)+' (m/s)')
    
    # ambient pressure for impulse computation
    Pa = 0.0
    print('Computing isentropic expansion')
    pp = Pt
    
    P = []; R = []; T = [];
    h_eq = []; u_eq = []; Isp_eq = [];
    h_fr = []; R_fr = []; T_fr = []; u_fr = []; Isp_fr = [];
    
    imax = 90
    for i in range(imax):
        pp = pp*0.95
        # compute equilibrium isentrope
        gas.SPX = St,pp,gas.X        
        gas.equilibrate('SP')
        
        P.append(gas.P)
        R.append(gas.density)
        T.append(gas.T)
        h_eq.append(gas.enthalpy_mass)
        u_eq.append(np.sqrt(2*(Ht-h_eq[i])))
        Isp_eq.append((u_eq[i]+ (P[i] - Pa)/(u_eq[i]*R[i]))/9.81)
        # compute frozen impulse  (first call is so we get a good guess for entropy)
        gas1.TPX = T[i],pp,qc
        gas1.SPX = St,pp,qc
        h_fr.append(gas1.enthalpy_mass)
        R_fr.append(gas1.density)
        T_fr.append(gas1.T)
        u_fr.append(np.sqrt(2*(Ht-h_fr[i])))
        Isp_fr.append((u_fr[i]+ (P[i] - Pa)/(u_fr[i]*R_fr[i]))/9.81)

    I_fr.append(Isp_fr[-1])
    I_eq.append(Isp_eq[-1])
    print('final pressure '+str(P[-1])+' (Pa)')
    print('final specific impulse (eq) '+str(Isp_eq[-1])+' (s)')
    print('final velocity (eq) '+str(u_eq[-1])+' (m/s)')
    print('final specific impulse (fr) '+str(Isp_fr[-1])+' (s)')
    print('final velocity (fr) '+str(u_fr[-1])+' (m/s)')
    print('-------------------------------------------\n\n')


#
# Plot the impulses
#
plt.figure
fontsize = 12
plt.plot(he, I_fr, 'r')
plt.plot(he, I_eq, 'b')
plt.xlabel('helium mole fraction ',fontsize=fontsize);
plt.ylabel('Isp (s)',fontsize=fontsize);
plt.title('Warm gas thruster performance',fontsize=fontsize);
plt.legend(['Frozen','Equilibrium'])
plt.tick_params(labelsize=12)
plt.xlim(xmin=0)

plt.show()
